### ISSUE TITLE
Custom socket & protocol support?

### ISSUE THREAD

**Issue by Aciid** [serverless/serverless#735]
Inspired by #647 I'm interested in does / will Serverless / AWS IoT Lambda ever support Custom Sockets and decode / encode protocols.

Say you could have a TCP/UDP socket server that accepts everything and or requires certain custom authenication flow, which would allow you to refine the sent data and allow sending data back through the same TCP session.

This would furtherly and ultimately allow implementing legacy IoT devices "Device cloud" -devices. That have usually been developed using a certain platforms such as Digi's [Device cloud](http://www.digi.com/products/cloud/digi-device-cloud) etc. This would help developers and end-engineers help to graduate their devices off from 3rd-party clouds. Given that the end-points can usually be redirected to something like what you can build with Serverless and AWS IoT.

A common example would be just to dissect the incoming first handsake TCP -packet, that contains a few known bytes + UID after the header. Then the further communication will be data sent from the device after that you usually send your ACK that you have "received" and or commands.

**Comment by flomotlik** [comment]
Hi @Aciid , we're going to depend on AWS (and other cloud providers) to deliver this. We will integrate deeper with AWS and cloudformation in an upcoming release so you should have an easier time setting this up. Once IOT supports those extensions they should be available to serverless users as well.

I will close this issue for now as this is dependent on AWS implementing new features and we should have the general ability to go there if AWS supports it in the future.

**Comment by Aciid** [comment]
Hey @flomotlik , thanks for the in-depth answer. I've worked with serverless now for some 3 weeks and I've gotten around it pretty well and I think I understand the situation a little better now and how these different techonologies are related.

After reading a little deeper from AWS Lambda documentation it's noted that "Inbound connections are managed by lambda TCP/IP sockets are supported, ptrace restricted and TCP port 25 restricted" 

So I figured that this could still work from the viewpoint of nodejs on lambda. 
Have to see could it work on serverless, but for the time being ill try to come up with a definite concrete  "socket server" lambda deployment first and then with some suggestions here how this could fit in serverless 
( Surely custom socket protocols are not directly related to AWS IoT as it's using MQTT. But still widely used around the industry. )
